plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+		g_prefix_error (error, "failed to send fw update start cmd: ");
+		return FALSE;
+	}
+
+	/* send data */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	chunks = fu_chunk_array_new_from_bytes (fw,
+						0x00,	/* start addr */
+						0x00,	/* page_sz */
+						ISP_DATA_BLOCKSIZE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+			return FALSE;
+		if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+							   UC_FOREGROUND_SLAVE_ADDR,
+							   UC_FOREGROUND_ISP_DATA_OPCODE,
+							   fu_chunk_get_data (chk),
+							   fu_chunk_get_data_sz (chk),
+							   error)) {
+			g_prefix_error (error,
+					"failed to write @0x%04x: ",
+					fu_chunk_get_address (chk));
+			return FALSE;
+		}
+
+		/* update progress */
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len);
+	}
+
+	/* update finish command */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+		return FALSE;
+	write_buf[0] = ISP_CMD_FW_UPDATE_ISP_DONE;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   write_buf, 1,
+						   error)) {
+		g_prefix_error (error, "failed update finish cmd: ");
+		return FALSE;
+	}
+
+	/* exit background-fw mode */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_progress (device, 0);
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+		return FALSE;
+	write_buf[0] = ISP_CMD_FW_UPDATE_EXIT;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   write_buf, 1,
+						   error)) {
+		g_prefix_error (error, "exit background-fw mode: ");
+		return FALSE;
+	}
+
+	/* the device needs some time to restart with the new firmware before
+	* it can be queried again */
+	fu_device_sleep_with_progress (device, 20);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_open (FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (device);
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no parent device");
+		return FALSE;
+	}
+	return fu_device_open (parent, error);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_close (FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (device);
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no parent device");
+		return FALSE;
+	}
+	return fu_device_close (parent, error);
+}
+
+static void
+fu_rts54hub_rtd21xx_device_init (FuRts54hubRtd21xxDevice *self)
+{
+	fu_device_add_icon (FU_DEVICE (self), "video-display");
+	fu_device_set_protocol (FU_DEVICE (self), "com.realtek.rts54");
```

Comment:
this kinda seems like a new protocol compared to how the hub itself updates.  is this accurate?

Reply:
Agree. This should probably be `com.realtek.rts54.i2c` -- fixed.